### PR TITLE
Protobuf

### LINKED ISSUE #74: Instead of hanging connections on unknown codes, return an error and handle it accordingly
For example we could try to reestablish the connection / create a new one.

cc #42

### PR DESCRIPTION
**Pull request by main--**
WIP. Fixes #74.

### CODE REVIEW

**Review by jonas-wanninger — APPROVED** [review]
Great m8

**Comment by moritzuehling** [comment]
It is now hopefully in a working state. It does work for me, including all features I tested (Hotkeys, Clipboard)

**Comment by moritzuehling** [comment]
It is now in a working state with every feature I tested (Clipboard, Hotkeys, Ping, ReportBoot). Only suspend is untested.

**Comment by main--** [comment]
>	/usr/lib/mono/xbuild/14.0/bin/Microsoft.Common.targets:  warning : Reference 'Google.Protobuf, Version=3.3.0.0, Culture=neutral, PublicKeyToken=a7d26565bac4d604, processorArchitecture=MSIL' not resolved

Build broken because the Makefile does not restore nuget packages (we should just add a nuget build-dependency on arch).

The Makefile is also missing the section to update the C# proto definition. Would you mind adding this?

File: windows-gaming-driver/src/driver/clientpipe/mod.rs [hunk 1]
```diff
@@ -73,33 +74,17 @@ impl Clientpipe {
...
-                    }
-                    Some(ClipboardMessage::ContentsType(t)) => {
-                        unimplemented!();
-                    }
-                    Some(ClipboardMessage::ClipboardContents(buf)) => {
-                        debug!("windows responded with clipboard contents");
-                        controller.borrow_mut().respond_x11_clipboard(buf);
-                    }
+                    Some(ClipboardMessage::GrabClipboard(_)) => controller.borrow_mut().grab_x11_clipboard(),
+                    Some(ClipboardMessage::RequestClipboardContents(kind)) => match ClipboardType::from_i32(kind) {
```
Comment (oberien) [inline windows-gaming-driver/src/driver/clientpipe/mod.rs]: Why do we need `ClipboardType::from_i32` here? Shouldn't that already be serialized by prost?

Comment (main--) [reply windows-gaming-driver/src/driver/clientpipe/mod.rs]: No. The protobuf spec says that implementations must be able to read enum values they don't (yet) know about. Prost deals with this by always representing them as int and offering conversion functions to the enum.

File: windows-gaming-driver/src/driver/clipboard.rs [hunk 2]
```diff
@@ -137,11 +161,18 @@ impl X11Clipboard {
                     let event: &SelectionNotifyEvent = unsafe { xcb::cast_event(&event) };
                     if let Ok(reply) = xcb::get_property(
                         &self.connection, false, self.window,
-                        event.property(), self.atoms.utf8_string, 0, ::std::u32::MAX // FIXME reasonable buffer size
+                        event.property(), xcb::ATOM_ANY, 0, ::std::u32::MAX // FIXME reasonable buffer size
                     ).get_reply() {
```
Comment (oberien) [inline windows-gaming-driver/src/driver/clipboard.rs]: Could you please refactor this to `let res = ...; if let Ok(_) = res`?

**Review by oberien — DISMISSED** [review]
lgtm

**Review by oberien — APPROVED** [review]

**Pull request merged**